libV2/vault/index.js

```diff
@@ -0,0 +1,117 @@
+const assert = require('assert');
+const { vault } = require('./client');
+const metadata = require('../metadata');
+const errors = require('../errors');
+
+// Will be used for user keys
+// eslint-disable-next-line no-unused-vars
+async function translateResourceIds(level, resources, log) {
+    if (level === 'accounts') {
+        return vault.getCanonicalIds(resources, log);
+    }
+
+    return resources.map(resource => ({ resource, id: resource }));
+}
+
+
+async function authorizeAccountAccessKey(authInfo, level, resources, log) {
+    let authed = false;
+    let authedRes = [];
+
+    switch (level) {
+    // Account keys can only query metrics their own account metrics
+    // So we can short circuit the auth to ->
+    // Did they request their account? Then authorize ONLY their account
+    case 'accounts':
+        authed = resources.some(r => r === authInfo.getShortid());
+        authedRes = authed ? [{ resource: authInfo.getShortid(), id: authInfo.getCanonicalID() }] : [];
+        break;
+
+    // Account keys are allowed access to any of their child users metrics
+    case 'users': {
+        let users;
+        try {
+            users = await vault.getUsersById(resources, log.logger);
+        } catch (error) {
+            log.error('failed to fetch user', { error });
+            throw errors.AccessDenied;
+        }
+        authedRes = users
+            .filter(user => user.parentId === authInfo.getShortid())
+            .map(user => ({ resource: user.id, id: user.id }));
+        authed = authedRes.length !== 0;
+        break;
+    }
+
+    // Accounts are only allowed access if they are the owner of the bucket
+    case 'buckets': {
+        let buckets;
+        try {
+            buckets = await Promise.all(
+                resources.map(bucket => metadata.getBucket(bucket)),
+            );
+        } catch (error) {
+            log.error('failed to fetch metadata for bucket', { error });
```

Comment:
It could be more precise to log right after `metadata.getBucket` returns, to log which bucket got an error